storage/remote/wal_watcher.go

```diff
@@ -326,9 +327,17 @@ func (w *WALWatcher) watch(segmentNum int, tail bool) error {
 		case <-checkpointTicker.C:
 			// Periodically check if there is a new checkpoint so we can garbage
 			// collect labels. As this is considered an optimisation, we ignore
-			// errors during checkpoint processing.
-			if err := w.garbageCollectSeries(segmentNum); err != nil {
-				level.Warn(w.logger).Log("msg", "error process checkpoint", "err", err)
+			// errors during checkpoint processing, and do the process asynchronously.
```

Comment:
Can you leave another comment, something like `locking of the series labels within garbageCollectSeries isn't done until SeriesReset is called, so we can still process the current WAL segment while reading the checkpoint`.

Reply:
Hmmm, where do you think the best place for a comment would be? Perhaps on the `writeTo` interface and say something about `SeriesReset` might be called concurrently with the other two?

Reply:
I added a comment here, and at the `WriteTo` interface noting that the functions might be called concurrently. Note that garbage collection calls both `SeriesReset` and `StoreSeries` so there will be lots of concurrent calls happening.